tsinfer/algorithm.py

```diff
@@ -1118,3 +1119,166 @@ def run_traceback(self, start, end, match):
             parent[j] = e.parent
 
         return left, right, parent
+
+
+def merge_overlapping_ancestors(start, end, time):
+    # Merge overlapping, same-time ancestors. We do this by scanning along a single
+    # time epoch from left to right, detecting breaks.
+    sort_indices = np.lexsort((start, time))
+    start = start[sort_indices]
+    end = end[sort_indices]
+    time = time[sort_indices]
+    old_indexes = {}
+    # For efficiency, pre-allcate the output arrays to the maximum possible size.
+    new_start = np.full_like(start, -1)
+    new_end = np.full_like(end, -1)
+    new_time = np.full_like(time, -1)
+
+    i = 0
+    new_index_pos = 0
+    while i < len(start):
+        j = i + 1
+        group_overlap = [i]
+        max_right = end[i]
+        # While we're in the same time epoch, and the next ancestor
+        # overlaps with the group, add this ancestor to the group.
+        while j < len(start) and time[j] == time[i] and start[j] < max_right:
+            max_right = max(max_right, end[j])
+            group_overlap.append(j)
+            j += 1
+
+        # Emit the found group
+        old_indexes[new_index_pos] = group_overlap
+        new_start[new_index_pos] = start[i]
+        new_end[new_index_pos] = max_right
+        new_time[new_index_pos] = time[i]
+        new_index_pos += 1
+        i = j
+    # Trim the output arrays to the actual size.
+    new_start = new_start[:new_index_pos]
+    new_end = new_end[:new_index_pos]
+    new_time = new_time[:new_index_pos]
+    return new_start, new_end, new_time, old_indexes, sort_indices
+
+
+@numba.njit
+def run_linesweep(event_times, event_index, event_type, new_time):
+    # Run the linesweep over the ancestor start-stop events,
+    # building up the dependency graph as a count of dependencies for each ancestor,
+    # and a list of dependant children for each ancestor.
+    n = len(new_time)
+
+    # numba really likes to know the type of the list elements, so we tell it by adding
+    # a dummy element to the list and then popping it off.
+    # `active` is the list of ancestors that overlap with the current linesweep position.
+    active = [-1]
```

Comment:
FYI you can also do this by explicitly creating a TypedList (I think it's called). This is fine though

Reply:
Yeah, I was put off by the "Experimental Feature" warning in the numba docs.